presc/spatial_distribution/categorical_distance.py

```diff
@@ -0,0 +1,509 @@
+
+        if d1_attributes == d2_attributes:
+            intersection = dpoint1 == dpoint2
+            common_attributes = list(dpoint1[intersection].index)
+            goodall2_score = 0
+            for attribute in common_attributes:
+                attribute_score = 0
+                counts = self._counts_per_attribute[attribute]
+
+                for count in counts:
+                    attribute_score = self.goodall_frequency(count) + attribute_score
+                    pass
+
+                goodall2_score = (1 - attribute_score) + goodall2_score
+
+            return goodall2_score / len(d1_attributes)
+        else:
+            # Raise exeption points do not have the same attributes
+            pass
+        pass
+
+    def goodall3(self, dpoint1, dpoint2):
+        """Computes the goodall3 similarity measurment for categorical data, see paper by
+        Varun, Shyam, Vipin in the bibliography reference for reference"""
+
+        d1_attributes = list(dpoint1.index)
+        d2_attributes = list(dpoint2.index)
+        goodall3_score = 0
+        if d1_attributes == d2_attributes:
+            for attribute in d1_attributes:
+                attribute_score = 0
+                value_point1 = dpoint1[attribute]
+                value_point2 = dpoint2[attribute]
+                if value_point1 == value_point2:
+                    count = self._counts_per_attribute[attribute][value_point1]
+
+                    attribute_score = self.goodall_frequency(count)
+
+                    goodall3_score = (1 - attribute_score) + goodall3_score
+
+            return goodall3_score / len(d1_attributes)
+        else:
+            # write exception
+            pass
+
+        pass
+
+    def overlap(self, dpoint1, dpoint2):
+        """Computes the overlap similarity measure for categorical data"""
+
+        if list(dpoint1.index) == list(dpoint2.index):
+            overlap_score = len(set(dpoint1) & set(dpoint2)) / len(dpoint1)
+            return overlap_score
+        else:
+            # write exception
+
+            pass
+
+    def lin(self, dpoint1, dpoint2):
+        d1_attributes = list(dpoint1.index)
+        d2_attributes = list(dpoint2.index)
+
+        if d1_attributes == d2_attributes:
+            lin = 0
+            weight = self.lin_weight(dpoint1, dpoint2)
+            for attribute in d1_attributes:
+                frequency_dpoint1 = self.empirical_frequency(
+                    dpoint1[attribute], attribute
+                )
+                frequency_dpoint2 = self.empirical_frequency(
+                    dpoint2[attribute], attribute
+                )
+                if dpoint1[attribute] == dpoint2[attribute]:
+                    lin = lin + 2 * np.log(frequency_dpoint1)
+                else:
+                    lin = lin + 2 * np.log(frequency_dpoint1 + frequency_dpoint2)
+        return weight * lin
+
+    def lin_weight(self, dpoint1, dpoint2):
+        d1_attributes = list(dpoint1.index)
```

Comment:
I wanted to keep it separate because I thought it cluttered the code and having it separately was more neat. I can integrate it if you think that's easier to follow

Reply:
I agree in general breaking computations into smaller functions is a good practice, but here it seems unnecessary to recompute all the empirical frequencies. I would just accumulate the `weight_denominator` inside the `for` loop in `lin()`.